### DISCUSSION TITLE
Using shadow to run an AWS SQS Client

### DISCUSSION THREAD

**Discussion by akumar1512** [shadow/shadow#2786] [Q&A] [answered]
**Issue**
Shadow does not log any output for calls it makes to run an SQS client even if the simulation ends successfully. However I can see some apache setup logs in the stdout file which is expected before starting application instructions but the main application requests logs seems missing.

**Aboout App and Setup:**
So, this is a simple SQS app that I am trying to run on Shadow. It initializes an `AWSCredentialProvider`  and initiates `AmazonSQSClientBuilder` based on the credentials generated.
The app then makes request to `SQS server` with a URL and queue name `(https://sqs.us-west-2.amazonaws.com/<accountID>/<queueName>)`. Post which there are some test read/writes on the queue.

I also tried with one more setup where I run the same SQS Client with a management app on Shadow that is supposed to filters all the requests made by SQS client (by printing all the hostname to ip resolutions for the run). `(Query #3)`

**Queries**

1. If shadow is able to run the client successfully, how is it able to emulate the credentials, initialize queue and make SQS calls. Does shadow support mock implementation of these services `(MOTO/localstack versions of SQS, SNS, etc)` ?

2. I also tried looking into the syscalls made by the application and I can see one request to shadow_hostname_to_addr_ipv4 where requested `name` is equal to `host_getName(_syscallhandler_getHost(sys))` and thus returns the same address - 
`00:00:03.125348 [2849:shadow-worker] 00:00:01.027117350 [INFO] [sqs-client:11.0.0.2] [shadow.c:76] [syscallhandler_shadow_hostname_to_addr_ipv4] Found address sqs-client-11.0.0.2 (eth,mac=4) for name sqs-client`. 
I was expecting a SQS URL's hostname resolution here.

3. I have built a management app to run on shadow that intercepts all the http requests to itself. This app for my test codes shows correct behavior (i.e I am able to see the requests made by my test code) by listing all the calls made by the test app. I cannot see the same for the SQS client when run alongside the management app.

I have already tried both trace/debug output options but with no luck. Any suggestions on why the log outputs are not visible and what can be done to observe them. Happy to provide any further inputs.

**Comment by stevenengler** [comment, marked as answer]
Can you add more details about what you're trying to do, and how? I'm not familiar with any of the Amazon stuff. SQS seems to be an Amazon cloud thing which you're trying to access from a program running in Shadow? Normally this wouldn't be possible since Shadow emulates the network, and doesn't allow access to a real external network. But you've been working on disabling interposition I think based on your previous questions? So I'm guessing that you're trying to access this external SQS service when interposition is disabled?

> Shadow does not log any output for calls it makes to run an SQS client even if the simulation ends successfully.

> I have already tried both trace/debug output options but with no luck. Any suggestions on why the log outputs are not visible and what can be done to observe them. Happy to provide any further inputs.

Can you expand on this? What output are you looking for, and where do you expect it to be? Something that's written by the program to stdout or stderr? Or to a log file? Is interposition enabled when it's doing the logging? Have you checked the `shimlog` file?

> If shadow is able to run the client successfully, how is it able to emulate the credentials, initialize queue and make SQS calls.

I just want to clarify what it means for the Shadow simulation to be "successful". Shadow considers the simulation to be successful if no processes exited with a non-zero exit code, **except** if the process was killed at the end of the simulation. At the end of the simulation, shadow sends a SIGKILL to all processes that are still running. These processes will be killed, but Shadow still considers this as "successful", even though they exited due to a signal. We're not completely happy with this behaviour, and it may change in the future. You can read more at https://github.com/shadow/shadow/discussions/2034.

So this all means that if your SQS client does something that blocks, and it blocks until the end of the simulation, shadow will send a SIGKILL at the end of the simulation but still say that the simulation was successful. For example if your SQS client tries to connect using a blocking socket to a server that isn't listening, that `connect()` call might block until the end of the simulation. So you should check that your program isn't blocked on anything. This might be the reason you don't see any logs, because it's actually blocked on a syscall and never continues past that.

> Does shadow support mock implementation of these services (MOTO/localstack versions of SQS, SNS, etc) ?

I don't know what those are, so probably not :)

**Reply by akumar1512** [reply]
Thanks for getting back on this. 

> Can you add more details about what you're trying to do, and how?

Just a bit of background. This is with regard to the discussion we had in #2689 . Implementing what Rob suggested, I aim is to have a management app alongside a service that manages out-of-shadow requests to larger federation. For this, any http/https requests made by the service is routed to the management app. The management app then disables interposition to relay request to external server. I have validated this behavior using toy python http server examples and is able to observe the requests made by the service in management app's stdout.

I wanted to test same behavior now running a simple SQS client service (essentially making a normal call to SQS server outside shadow) alongside the management app. The expected behavior is to observe the requests made by the SQS client on management app logs.

> if your SQS client does something that blocks, and it blocks until the end of the simulation, shadow will send a SIGKILL at the end of the simulation but still say that the simulation was successful

I can relate this to my issue since these services initializes some security checks before making the actual SQS request. One of them is on a specific port on localhost. The usual behavior of this SQS service outside shadow is that it expects a mocked response from the port before proceeding to make SQS request.

I think the service is blocked here as I can see in my logs -

```
00:00:00.113582 [2849:shadow-worker] 00:00:01.000031020 [WARN] [sqs-client:11.0.0.2] [syscall_handler.c:552] [syscallhandler_make_syscall] Detected unsupported syscall 435 called from thread 1000 in process sqs-client.java.1000 on host sqs-client
00:00:00.113584 [2849:shadow-worker] 00:00:01.000031020 [ERROR] [sqs-client:11.0.0.2] [syscall_handler.c:556] [syscallhandler_make_syscall] Returning error 38 (ENOSYS) for unsupported syscall 435, which may result in unusual behavior
00:00:00.114889 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1000): handling syscall 56 clone
00:00:00.123001 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1000): handling syscall 14 rt_sigprocmask
00:00:00.123037 [2849:shadow-worker] 00:00:01.000033020 [DEBUG] [sqs-client:11.0.0.2] [process.rs:589] [shadow_rs::host::process] process 'sqs-client.java.1000' done continuing
00:00:00.123038 [2849:shadow-worker] 00:00:01.000033020 [DEBUG] [sqs-client:11.0.0.2] [process.rs:1150] [shadow_rs::host::process] thread 1000 in process 'sqs-client.java.1000' still running, but blocked
00:00:00.123053 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1001): handling syscall 1002 shadow_get_shm_blk
00:00:00.123055 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [shadow.c:123] [syscallhandler_shadow_get_shm_blk] handling shadow_get_shm_blk syscall
00:00:00.123105 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1001): handling syscall 334 rseq 
00:00:00.123149 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1001): handling syscall 273 set_robust_list 
00:00:00.123151 [2849:shadow-worker] 00:00:01.000033020 [DEBUG] [sqs-client:11.0.0.2] [futex.c:221] [syscallhandler_set_robust_list] set_robust_list was called but we don't yet support it 
00:00:00.123153 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:223] [_syscallhandler_post_syscall] Syscall didn't complete successfully; discarding plugin ptrs without writing back.
00:00:00.123202 [2849:shadow-worker] 00:00:01.000035020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1001): handling syscall 14 rt_sigprocmask
```
It seems like the [GetQueueUrlRequest](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sqs/model/GetQueueUrlRequest.html) is making some syscalls that are not supported. Can you share your comments?

How do I configure a `port on the localhost in shadow namespace` to send specific response to running service when called?
Is failing `set_robust_list` the reason for the blocking app?

>  Is interposition enabled when it's doing the logging?
I am presently testing everything related to the SQS service with interposition-enabled.

Rest of logs (repeating) -
```
00:00:04.544530 [2849:shadow-worker] 00:00:27.175617040 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1010): handling syscall 202 futex (previously BLOCKed)
00:00:04.544531 [2849:shadow-worker] 00:00:27.175617040 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:223] [_syscallhandler_post_syscall] Syscall didn't complete successfully; discarding plugin ptrs without writing back.
00:00:04.544576 [2849:shadow-worker] 00:00:27.175617040 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1010): handling syscall 202 futex
00:00:04.544621 [2849:shadow-worker] 00:00:27.175619040 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1010): handling syscall 202 futex
00:00:04.544622 [2849:shadow-worker] 00:00:27.175619040 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:223] [_syscallhandler_post_syscall] Syscall didn't complete successfully; discarding plugin ptrs without writing back.
00:00:04.544624 [2849:shadow-worker] 00:00:27.175619040 [DEBUG] [sqs-client:11.0.0.2] [process.rs:589] [shadow_rs::host::process] process 'sqs-client.java.1000' done continuing
00:00:04.544624 [2849:shadow-worker] 00:00:27.175619040 [DEBUG] [sqs-client:11.0.0.2] [process.rs:1150] [shadow_rs::host::process] thread 1010 in process 'sqs-client.java.1000' still running, but blocked
```

**Reply by stevenengler** [reply]
> I can relate this to my issue since these services initializes some security checks before making the actual SQS request. One of them is on a specific port on localhost. The usual behavior of this SQS service outside shadow is that it expects a mocked response from the port before proceeding to make SQS request.

> How do I configure a `port on the localhost in shadow namespace` to send specific response to running service when called?

Can you run this service inside of Shadow? If you can run this as part of the Shadow simulation, then communication over Shadow's emulated localhost should work fine. If it needs to run outside of Shadow, then you'll probably need to use your management app to proxy the traffic between Shadow and localhost outside of Shadow. Or if you can't run the service inside of Shadow, can you write a custom mock service that returns this mock response and run it in Shadow?

> Is failing `set_robust_list` the reason for the blocking app?

It's unlikely that this is an issue. My understanding is that `get_robust_list` and `set_robust_list` are useful when a thread exits or execs without releasing a futex, but Shadow doesn't yet support exec, and it's rare that a thread exits cleanly without cleaning up any futexes. Most applications that use glibc will make calls to `get_robust_list` and we haven't yet found any applications where returning `ENOSYS` causes an issue.

> ```
> 00:00:00.113582 [2849:shadow-worker] 00:00:01.000031020 [WARN] [sqs-client:11.0.0.2] [syscall_handler.c:552] [syscallhandler_make_syscall] Detected unsupported syscall 435 called from thread 1000 in process sqs-client.java.1000 on host sqs-client
> 00:00:00.114889 [2849:shadow-worker] 00:00:01.000033020 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1000): handling syscall 56 clone
> ```

Syscall 435 is `clone3()`, which should be fine since glibc will fall back to `clone()` which Shadow does support.

It looks like you changed the `SYSCALL_HANDLER_PRE` log lines from "trace" to "info". Can you also change the `SYSCALL_HANDLER_POST` lines to "info"? It might show better where it's blocking. For example, the `SYSCALL_HANDLER_POST` should show something like:

```
[TRACE] [testnode:11.0.0.1] [syscall_handler.c:167] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(test_connect,pid=1000): handling syscall 42 connect
[...]
[TRACE] [testnode:11.0.0.1] [syscall_handler.c:204] [_syscallhandler_post_syscall] SYSCALL_HANDLER_POST(test_connect,pid=1000): syscall 42 connect result: state=BLOCK val=n/a(n/a)
```

**Reply by akumar1512** [reply]
Thanks for your reply. `SYSCALL_HANDLER_POST` shows `connect` getting blocked. What is strange is that I have the authentication server running on localhost at port `3000`, still why it is getting blocked. Pasting some relevant logs -

While startup -

```
00:00:00.027921 [13462:shadow] n/a [DEBUG] [n/a] [network_interface.c:398] [networkinterface_new] bringing up network interface 'gatewayApp' at '127.0.0.1' using queuing discipline fifo
00:00:00.028000 [13462:shadow] n/a [DEBUG] [n/a] [network_interface.c:398] [networkinterface_new] bringing up network interface 'gatewayApp' at '11.0.0.1' using queuing discipline fifo
```

Connection to mocked authentication request logs -

```
00:00:14.184654 [13476:shadow-worker] 00:00:01.032428450 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:160] [_syscallhandler_pre_syscall] SYSCALL_HANDLER_PRE(java,pid=1001): handling syscall 42 connect
00:00:14.184679 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [socket.c:900] [syscallhandler_connect] trying to connect on socket 67900:00:14.184704 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [socket.c:244] [_syscallhandler_bindHelper] trying to bind to inet address 127.0.0.1:0 on socket 0x7f71b0147d20 with peer 127.0.0.1:3000
00:00:14.184724 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [socket.c:263] [_syscallhandler_bindHelper] binding to generated ephemeral port 31308
00:00:14.184739 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [descriptor.c:74] [legacyfile_ref] Descriptor 0x7f71b0147d20 strong_ref++ to 2 (weak_ref=0)
00:00:14.184747 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [network_interface.c:119] [networkinterface_associate] associated socket key TCP|16777343:19578|16777343:55559
00:00:14.184753 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [tcp.c:929] [_tcp_sendControlPacket] 127.0.0.1:31308 (descriptor 0x7f71b0147d20) <-> 127.0.0.1:3000: sending response control packet now
00:00:14.184775 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [SND_CREATED] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:3000 seq=1 ack=0 sack=NA window=0 bytes=0 header=SYN tsval=0 tsechoreply=0 status=SND_CREATED
00:00:14.184791 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [SND_TCP_ENQUEUE_THROTTLED] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:3000 seq=1 ack=0 sack=NA window=0 bytes=0 header=SYN tsval=0 tsechoreply=0 status=SND_CREATED,SND_TCP_ENQUEUE_THROTTLED
00:00:14.184805 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [SND_SOCKET_BUFFERED] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:3000 seq=1 ack=0 sack=NA window=0 bytes=0 header=SYN tsval=0 tsechoreply=0 status=SND_CREATED,SND_TCP_ENQUEUE_THROTTLED,SND_SOCKET_BUFFERED
00:00:14.184819 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [descriptor.c:74] [legacyfile_ref] Descriptor 0x7f71b0147d20 strong_ref++ to 3 (weak_ref=0)
00:00:14.184835 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [descriptor.c:198] [_legacyfile_handleStatusChange] Status changed on desc 0x7f71b0147d20, from NONE to WRITEABLE
00:00:14.184844 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [tcp.c:1626] [_tcp_connectToPeer] 127.0.0.1:31308 (descriptor 0x7f71b0147d20) <-> 127.0.0.1:3000: user initiated connection
00:00:14.184849 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [tcp.c:647] [_tcp_setState] 127.0.0.1:31308 (descriptor 0x7f71b0147d20) <-> 127.0.0.1:3000: moved from TCP state 'TCPS_CLOSED' to 'TCPS_SYNSENT'
00:00:14.184854 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [descriptor.c:74] [legacyfile_ref] Descriptor 0x7f71b0147d20 strong_ref++ to 4 (weak_ref=0)
00:00:14.184865 [13476:shadow-worker] 00:00:01.032428450 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:198] [_syscallhandler_post_syscall] SYSCALL_HANDLER_POST(java,pid=1001): syscall 42 connect result: state=BLOCK val=n/a(n/a)

.....

00:00:14.184997 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [SND_INTERFACE_SENT] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:3000 seq=1 ack=1 sack=NA window=119 bytes=0 header=SYN tsval=1032428450 tsechoreply=0 status=SND_CREATED,SND_TCP_ENQUEUE_THROTTLED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT
00:00:14.185043 [13476:shadow-worker] 00:00:01.032428450 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [RELAY_FORWARDED] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:3000 seq=1 ack=1 sack=NA window=119 bytes=0 header=SYN tsval=1032428450 tsechoreply=0 status=SND_CREATED,SND_TCP_ENQUEUE_THROTTLED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT,RELAY_FORWARDED
00:00:14.185076 [13476:shadow-worker] 00:00:01.032428451 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [RCV_INTERFACE_RECEIVED] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:3000 seq=1 ack=1 sack=NA window=119 bytes=0 header=SYN tsval=1032428450 tsechoreply=0 status=SND_CREATED,SND_TCP_ENQUEUE_THROTTLED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT,RELAY_FORWARDED,RCV_INTERFACE_RECEIVED
00:00:14.185086 [13476:shadow-worker] 00:00:01.032428451 [TRACE] [sqs-client:11.0.0.2] [network_interface.c:190] [networkinterface_push] looking for socket associated with specific key TCP|16777343:55559|16777343:19578
00:00:14.185091 [13476:shadow-worker] 00:00:01.032428451 [TRACE] [sqs-client:11.0.0.2] [network_interface.c:198] [networkinterface_push] looking for socket associated with general key TCP|16777343:55559|0:0
00:00:14.185115 [13476:shadow-worker] 00:00:01.032428451 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [RCV_INTERFACE_DROPPED] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:3000 seq=1 ack=1 sack=NA window=119 bytes=0 header=SYN tsval=1032428450 tsechoreply=0 status=SND_CREATED,SND_TCP_ENQUEUE_THROTTLED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT,RELAY_FORWARDED,RCV_INTERFACE_RECEIVED,RCV_INTERFACE_DROPPED

....

00:00:14.185389 [13476:shadow-worker] 00:00:01.040379850 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:198] [_syscallhandler_post_syscall] SYSCALL_HANDLER_POST(java,pid=1010): syscall 202 futex result: state=BLOCK->DONE val=-110(Connection timed out)

....

00:00:14.789144 [13476:shadow-worker] 00:00:08.032428451 [TRACE] [sqs-client:11.0.0.2] [packet.c:729] [packet_addDeliveryStatus] [RCV_INTERFACE_DROPPED] packetID=1:0 127.0.0.1:31308 -> 127.0.0.1:2009 seq=1 ack=1 sack=NA window=119 bytes=0 header=SYN tsval=8032428450 tsechoreply=0 status=SND_CREATED,SND_TCP_ENQUEUE_THROTTLED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT,RELAY_FORWARDED,RCV_INTERFACE_RECEIVED,RCV_INTERFACE_DROPPED,SND_TCP_DEQUEUE_RETRANSMIT,SND_TCP_ENQUEUE_THROTTLED,SND_TCP_RETRANSMITTED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT,RELAY_FORWARDED,RCV_INTERFACE_RECEIVED,RCV_INTERFACE_DROPPED,SND_TCP_DEQUEUE_RETRANSMIT,SND_TCP_ENQUEUE_THROTTLED,SND_TCP_RETRANSMITTED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT,RELAY_FORWARDED,RCV_INTERFACE_RECEIVED,RCV_INTERFACE_DROPPED,SND_TCP_DEQUEUE_RETRANSMIT,SND_TCP_ENQUEUE_THROTTLED,SND_TCP_RETRANSMITTED,SND_SOCKET_BUFFERED,SND_TCP_ENQUEUE_RETRANSMIT,SND_INTERFACE_SENT,RELAY_FORWARDED,RCV_INTERFACE_RECEIVED,RCV_INTERFACE_DROPPED
```

> looking for socket associated with specific key.

Seems like it is not able to find the key for the associated socket and the receive interface is getting dropped instead of `RCV_SOCKET_PROCESSED`. The authentication app running here as the gateway app is configured to send response for requests on port `3000`. I expect some traffic on the interface. Do you also think it's blocked because it is not able to find the appropriate interface? Any suggest what else I can try here? Let me know if you need any other info from the logs.

**Reply by stevenengler** [reply]
Are you running the gateway and client on different hosts? From the logs it looks like the gateway app is running on 11.0.0.1 and the SQS client is running on 11.0.0.2. If so, then they can't communicate over localhost since they are on different hosts.

Either the gateway app and SQS client need to run on the same host, or you need to have the gateway app listen on 0.0.0.0:3000 and have the SQS client connect to 11.0.0.1:3000 instead.

**Reply by akumar1512** [reply]
> Are you running the gateway and client on different hosts?

I think this makes sense why shadow is not able to find the interface.  Even the `shadow_hostname_to_addr_ipv4 `syscall results into SQS service's own address.

```
00:00:10.690713 [24044:shadow-worker] 00:00:01.027174530 [INFO] [sqs-client:11.0.0.2] [shadow.c:68] [syscallhandler_shadow_hostname_to_addr_ipv4] Using default address for my own hostname sqs-client
00:00:10.690717 [24044:shadow-worker] 00:00:01.027174530 [INFO] [sqs-client:11.0.0.2] [shadow.c:76] [syscallhandler_shadow_hostname_to_addr_ipv4] Found address sqs-client-11.0.0.2 (eth,mac=4) for name sqs-client
00:00:10.690729 [24044:shadow-worker] 00:00:01.027174530 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:198] [_syscallhandler_post_syscall] SYSCALL_HANDLER_POST(java,pid=1001): syscall 1003 shadow_hostname_to_addr_ipv4 result: state=DONE val=0(n/a)
```

I tried one of your suggestion. Since I cannot change anything in the SQS service, I changed the gateway app to listen on `0.0.0.0:3000`. The issue still persists. 

Is there any way I can run two processes on same host using shadow config options?

I see the [doc](https://shadow.github.io/docs/guide/shadow_config_spec.html#hostshostnameprocessesargs) mentioning `hosts.<hostname>.processes[*].args`
Does this mean we have an option of having multiple process on same host?

For example:
```

hosts:
  nt_app_play_pause:
    network_node_id: 0
    options:
      log_level: trace
    processes[1]:
      - path: path
        args: arg
        start_time: 0s
    processes[2]:
      - path: path
        args: arg
        start_time: 1s
```

**Reply by stevenengler** [reply]
> Is there any way I can run two processes on same host using shadow config options?
>
> I see the doc mentioning `hosts.<hostname>.processes[*].args`
> Does this mean we have an option of having multiple process on same host?

Yeah it's just a yaml list of processes to run on the host:

```yaml
hosts:
  nt_app_play_pause:
    network_node_id: 0
    options:
      log_level: trace
    processes:
      - path: path
        args: arg
        start_time: 0s
      - path: path
        args: arg
        start_time: 1s
```

See the [etcd example](https://github.com/shadow/shadow/blob/f397d414a6880e8d95d2f44cc6eb9b56fc0ee24a/examples/etcd/shadow.yaml) and [tor-minimal test](https://github.com/shadow/shadow/blob/f397d414a6880e8d95d2f44cc6eb9b56fc0ee24a/src/test/tor/minimal/tor-minimal.yaml) for other examples.

**Reply by akumar1512** [reply]
Thanks @stevenengler for your prompt replies. Happy to be a step closer to my goal :)

```
00:00:14.197426 [24080:shadow-worker] 00:00:01.032422332 [INFO] [sqs-client:11.0.0.2] [syscall_handler.c:198] [_syscallhandler_post_syscall] SYSCALL_HANDLER_POST(java,pid=1002): syscall 42 connect result: state=BLOCK->DONE val=0(n/a)
```
